larq/quantized_variable.py

```diff
@@ -27,6 +30,7 @@ def __init__(
         variable: tf.Variable,
         quantizer: Optional[Quantizer] = None,
         precision: Optional[int] = None,
+        op: Optional[tf.Operation] = UNSPECIFIED,
```

Comment:
I'm surprised this passes type checking, why does `UNSPECIFIED` (which is equal to `object()`) satisfy `Optional[tf.Operation]`?

Is there a problem using `None` as the default value here, or can `None` be interpreted as a meaningful `tf.Operation`?

Reply:
Ah I see what the issue is, you have to know whether or not to forward the `op` of the underlying variable or not, which is why you use a sentinel object rather than just `None`?

In that case maybe we could do something [like this](https://github.com/larq/zookeeper/blob/a7b1ddb8e054d9b0931278a176c0e872792841c7/zookeeper/core/utils.py#L10-L16) and define a class of the sentinel object, so that it can be referred to in the type hint rather than using `Optional`?

Reply:
> Ah I see what the issue is, you have to know whether or not to forward the op of the underlying variable or not, which is why you use a sentinel object rather than just None?

Correct, I need to distinguish between `None` and unset. 

> In that case maybe we could do something like this and define a class of the sentinel object, so that it can be referred to in the type hint rather than using Optional?

I'd still need to use `Optional` or `Union[tf.Operation, None]` in both cases since `op` is explicitely allowed to be `None`. I mainly opted for this solution since it is widely used in the TensorFlow codebase.

> I'm surprised this passes type checking, why does `UNSPECIFIED` (which is equal to object()) satisfy `Optional[tf.Operation]`?

Yeah, not sure if this is a bug or expected behaviour. Maybe it will implicitely make a union of the specified type and the default value?

I have the feeling that `pytype` is simply broken, it doesn't even throw an error for me if I add the following to this file:
```python
def foo(i: int):
    return i

foo("3.0") # this is clearly wrong
```